docs/source/distributed.pipelining.rst

```diff
@@ -4,182 +4,310 @@
+Note that the above code needs to be launched for each worker, thus we use a
+launcher service to launch multiple processes:
 
 .. code-block:: bash
 
-  torchrun --nproc_per_node=3 example.py
+  torchrun --nproc_per_node=2 example.py
+
+
+Hugging Face Examples
+*********************
+
+In the `PiPPy <https://github.com/pytorch/PiPPy>`_ repo where this package was
+original created, we kept examples based on unmodified Hugging Face models.
+See the `examples/huggingface
+<https://github.com/pytorch/PiPPy/tree/main/examples/huggingface>`_ directory.
+
+Examples include:
+
+* `GPT2 <https://github.com/pytorch/PiPPy/tree/main/examples/huggingface/pippy_gpt2.py>`_
+* `Llama <https://github.com/pytorch/PiPPy/tree/main/examples/llama>`_
+
+
+Technical Deep Dive
+*******************
+
+How does the ``pipeline`` API split a model?
+============================================
+
+First, the ``pipeline`` API turns our model into a directed acyclic graph (DAG)
+by tracing the model.  It traces the model using ``torch.export`` -- a PyTorch 2
+full-graph capturing tool.
+
+Then, it groups together the **operations** and **parameters** needed by a stage
+into a reconstructed submodule: ``submod_0``, ``submod_1``, ...
+
+Different from conventional submodule access methods like ``Module.children()``,
+the ``pipeline`` API does not only cut the module structure of your model, but
+also the **``forward``** function of your model.
+
+This is necessary because model structure like ``Module.children()`` merely
+captures information during ``Module.__init__()``, and does not capture any
+information about ``Module.forward()``. Said differently, ``Module.children()``
+lacks information on the following aspects which are key to pipelininig:
+
+* Exectuion order of child modules in ``forward``
+* Activation flows between child modules
+* Whether there are any functional operators between child modules (for example,
+  ``relu`` or ``add`` operations would be be captured by ``Module.children()``).
+
+The ``pipeline`` API, on the contrary, makes sure that the ``forward`` behavior
+is truly preserved. It also captures the activation flow between the partitions,
+helping the distributed runtime to make correct send/receive calls without human
+intervention.
+
+Another flexibility of the ``pipeline`` API is that split points can be at
+arbitrary hierarchy of your model. In the split partitions, the original model
+hierarchy related to that partition will be reconstructed at no cost of yours.
+At a result, fully-qualified names (FQNs) pointing to a submodule or parameter
+would be still valid, and services that relies on FQNs (such as FSDP, TP or
+checkpointing) can still run with your partitioned modules at almost zero code
+change.
+
```

Comment:
what do you think about adding a section about common issues with the tracer?

some issues are just bugs in the tracer and we should fix the tracer and not mention those here.

but other issues are more long-living and we rely on workaround such as 'configure the tracer to not trace into that module, or something' - we should probably give some guidance on that kind of thing?